File: packages/eep-middleware-python/eep_middleware/db/in_memory.py
from __future__ import annotations

from eep_middleware.adapters import SubscriptionRecord


class InMemoryDBAdapter:
    def __init__(self) -> None:
        self._items: dict[str, SubscriptionRecord] = {}

    async def save_subscription(self, subscription: SubscriptionRecord) -> None:
        self._items[subscription.subscription_id] = subscription

    async def get_subscription(self, subscription_id: str) -> SubscriptionRecord | None:
        return self._items.get(subscription_id)

    async def list_subscriptions(self) -> list[SubscriptionRecord]:
        return list(self._items.values())

File: packages/@eep-dev/middleware/src/db/in-memory.ts
import type { DBAdapter, SubscriptionRecord, SubscriptionUpdate } from "../core/request-handler.js";

export class InMemoryDBAdapter implements DBAdapter {
  private readonly records = new Map<string, SubscriptionRecord>();

  async saveSubscription(subscription: SubscriptionRecord): Promise<void> {
    this.records.set(subscription.subscription_id, subscription);
  }

  async getSubscription(subscriptionId: string): Promise<SubscriptionRecord | null> {
    return this.records.get(subscriptionId) ?? null;
  }

  async listSubscriptions(): Promise<SubscriptionRecord[]> {
    return Array.from(this.records.values());
  }

  async updateSubscription(subscriptionId: string, updates: SubscriptionUpdate): Promise<void> {
    const existing = this.records.get(subscriptionId);
    if (!existing) {
      return;
    }
    this.records.set(subscriptionId, { ...existing, ...updates });
  }

  async deleteSubscription(subscriptionId: string): Promise<boolean> {
    return this.records.delete(subscriptionId);
  }
}
